whoc/design_tools/wake_steering_design.py

```diff
@@ -0,0 +1,668 @@
+        offsets_stacked.shape[1]
+    )
+
+    # Add 360 to end, if full wind rose and wraps
+    if len(wind_directions) == 1:
+        raise ValueError("Cannot compute hysteresis regions for single wind direction.")
+    wd_steps = wind_directions[1:]-wind_directions[:-1]
+    if ((wind_directions[0] - wd_steps[0] < 0)
+        & (wind_directions[-1] + wd_steps[-1] >= 360)
+    ):
+        offsets = np.concatenate((offsets, offsets[0:1, :, :, :]), axis=0)
+        wind_directions = np.concatenate((wind_directions, [wind_directions[-1] + wd_steps[-1]]))
+        wd_steps = wind_directions[1:]-wind_directions[:-1]
+    wd_centers = wind_directions[:-1] + 0.5 * wd_steps
+
+    # Define function that identifies hysteresis zones
+    jump_threshold = yaw_rate_threshold*wd_steps[:,None,None,None]
+    jump_idx = np.argwhere(np.abs(np.diff(offsets, axis=0)) >= jump_threshold)
+    # Drop information about ws, ti
+    jump_idx = np.unique(jump_idx[:, [0, 3]], axis=0)
+    # Convert to a per-turbine dictionary of switching wind directions
+    centers_dict = {}
+    for t in np.unique(jump_idx[:,1]):
+        centers_dict["T{:03d}".format(t)] = (
+            wd_centers[jump_idx[jump_idx[:,1] == t][:,0]]
+        )
+    if verbose:
+        print("Center wind directions for hysteresis, per turbine: {}".format(centers_dict))
+        print("Computing hysteresis regions.")
+
+    # Find hysteresis regions for each switching point
+    # Note: doesn't handle the (unlikely) case that there is a large jump without a sign change
+    hysteresis_dict = {}
+    for turbine_tag in centers_dict.keys():
+        hysteresis_wds = []
+        for wd_switch_point in centers_dict[turbine_tag]:
+            t = int(turbine_tag[1:])
+            # Create region of minimum width
+            lb = wrap_360(wd_switch_point - min_zone_width/2)
+            ub = wrap_360(wd_switch_point + min_zone_width/2)
+            hysteresis_wds.append((lb, ub))
+
+        # Consolidate regions
+        for _ in range(len(hysteresis_wds)): # Outer loop to handle multiple possible wrap merges
+            i_h = 0
+            while i_h < len(hysteresis_wds)-1:
+                if (hysteresis_wds[i_h][1] >= hysteresis_wds[i_h+1][0]
+                    and hysteresis_wds[i_h][0] <= hysteresis_wds[i_h+1][1]):
+                    # Merge regions
+                    hysteresis_wds[i_h] = (
+                        min(hysteresis_wds[i_h][0], hysteresis_wds[i_h+1][0]),
+                        max(hysteresis_wds[i_h][1], hysteresis_wds[i_h+1][1])
+                    )
+                    # Remove next region
+                    hysteresis_wds.pop(i_h+1)
+                else:
+                    i_h += 1
+            # Check the final region, which could be reversed
+            if ((hysteresis_wds[-1][0] > hysteresis_wds[-1][1])
+                or (hysteresis_wds[0][0] > hysteresis_wds[0][1])):
+                if hysteresis_wds[-1][1] > hysteresis_wds[0][0]:
```

Comment:
Is the objective here to consolidate the first and last regions if they overlap around 0 degrees? Trying some scenarios, here are some examples where the regions wouldn't get combined:
`hysteresis_wds = [(3, 7), (359, 3)]`
`hysteresis_wds = [(359, 1), (357, 359)]`

I think this could be changed to `hysteresis_wds[-1][1] >= hysteresis_wds[0][0]`